### PR TITLE
Log metric for DNS type

### PR DESCRIPTION
**Pull request by mcleinman**
## Description

As part of the session package of metrics, we log the DNS type.

I considered more or less copying the `getDNS` function, but instead went with an approach that uses that function and translates the output to what we want to log. (To help, I also turned our DNS IP addresses into constants.) I don't feel strongly about this approach, and am happy to discuss better methods here.

## Reference

https://mozilla-hub.atlassian.net/browse/VPN-4067

## Checklist
    
- [ ] My code follows the style guidelines for this project
- [ ] I have not added any packages that contain high risk or unknown licenses (GPL,  LGPL, MPL, etc. consult with DevOps if in question)
- [ ] I have performed a self review of my own code
- [ ] I have commented my code PARTICULARLY in hard to understand areas
- [ ] I have added thorough tests where needed

### CODE REVIEW

File: src/apps/vpn/controller.cpp [hunk 1]
```diff
@@ -519,6 +520,8 @@ void Controller::connected(const QString& pubkey,
     resetConnectedTime();
   }
 
+  mozilla::glean::settings::dns_type.set(DNSHelper::getDNSType());
```
Comment (bakulf) [inline src/apps/vpn/controller.cpp:523]: Not here. But in telemetry.cpp

File: src/apps/vpn/dnshelper.cpp [hunk 2]
```diff
@@ -49,22 +57,52 @@ QString DNSHelper::getDNS(const QString& fallback) {
...
 }
 
+// static
+QString DNSHelper::getDNSType() {
+  if (!Feature::get(Feature::Feature_customDNS)->isSupported()) {
+    return "NoCustomDNSAvailable";
+  }
+
+  const QString testFallback = "anything";
+  static QMap<QString, QString> dnsTypeMap{
```
Comment (bakulf) [inline src/apps/vpn/dnshelper.cpp:85]: All of this is dup code.

File: src/apps/vpn/dnshelper.cpp [hunk 3]
```diff
@@ -49,22 +57,52 @@ QString DNSHelper::getDNS(const QString& fallback) {
   }
 
   static QMap<int, QString> dnsMap{
```
Comment (bakulf) [inline src/apps/vpn/dnshelper.cpp:59]: Instead of doing all of this, change the map to be:

QMap<int, DNSData>. Where DNSData is a simple struct composed by IP address and telemetry name.

Then use the telemetry name for the glean event.

What do you think?

**Review by bakulf — CHANGES_REQUESTED** [review]
Instead of this, add a connect-callback in telemetry. And improve the use of QMap to cover telemetry event names.

File: src/apps/vpn/controller.cpp (hunk 1 above)
Comment (mcleinman) [reply src/apps/vpn/controller.cpp:523]: Per discussion a couple days ago, I believe this is appropriate. We're going to reserve `telemetry.cpp` for complex telemetry recordings, multiline ones, etc.

File: src/apps/vpn/dnshelper.cpp (hunk 3 above)
Comment (mcleinman) [reply src/apps/vpn/dnshelper.cpp]: I like that suggestion *a lot*. I just implemented this, let me know if you were thinking about something different.

**Review by bakulf — APPROVED** [review]
r+ but remove the BLOCK* consts.

File: qtglean/metrics.yaml [hunk 4]
```diff
@@ -24,6 +24,7 @@ performance:
     expires: never
     time_unit: microsecond
 
+settings:
```
Comment (brizental) [inline qtglean/metrics.yaml:27]: How is this change related?

File: src/apps/vpn/dnshelper.cpp [hunk 5]
```diff
@@ -19,10 +19,28 @@ namespace {
...
+const QString BLOCK_MALWARE_ADS_TRACKERS_DNS = "100.64.0.7";
+
 // static
 QString DNSHelper::getDNS(const QString& fallback) {
+  return getDNSDetails(fallback).ipAddress;
+}
+
+// static
+QString DNSHelper::getDNSType() {
+  return getDNSDetails("test").dnsType;
```
[the remark this answers is not in the document]
Comment (brizental) [reply src/apps/vpn/dnshelper.cpp:37]: Whatever is the fallback needs to be documented on the metrics.yaml file.

**Review by brizental — CHANGES_REQUESTED** [review]
Same comment about adding unit tests: please do if possible. If not, file a follow up to do it.

File: src/apps/vpn/controller.cpp [hunk 6]
```diff
@@ -519,6 +520,8 @@ void Controller::connected(const QString& pubkey,
     resetConnectedTime();
   }
 
+  mozilla::glean::session::dns_type.set(DNSHelper::getDNSType());
```
[the remark this answers is not in the document]
Comment (mcleinman) [reply src/apps/vpn/controller.cpp:523]: Activation timestamp is a `datetime` metric, which doesn't have the ability or extra keys - that's exclusive to `event`s and maybe another 1-2 types.

File: src/apps/vpn/dnshelper.cpp [hunk 7]
```diff
@@ -19,10 +19,28 @@ namespace {
 Logger logger("DNSHelper");
 }
 
+const QString BLOCK_ADS_DNS = "100.64.0.1";
+const QString BLOCK_TRACKERS_DNS = "100.64.0.2";
+const QString BLOCK_ADS_TRACKERS_DNS = "100.64.0.3";
+const QString BLOCK_MALWARE_DNS = "100.64.0.4";
+const QString BLOCK_MALWARE_ADS_DNS = "100.64.0.5";
+const QString BLOCK_MALWARE_TRACKERS_DNS = "100.64.0.6";
+const QString BLOCK_MALWARE_ADS_TRACKERS_DNS = "100.64.0.7";
```
[the remark this answers is not in the document]
Comment (mcleinman) [reply src/apps/vpn/dnshelper.cpp:28]: I think it's better to pull out these IP addresses as constants, which is why I left it.

File: qtglean/metrics.yaml [hunk 8]
```diff
@@ -38,3 +39,21 @@ performance:
     notification_emails:
       - mcleinman@mozilla.com
     expires: never
+
+session:
+  dns_type:
+    type: string
+    description: |
+      The type of DNS a user has active.
+      (AdBlock | AntiTracking | AdBlockAndAntiTracking | Custom | Normal)
```
[the remark this answers is not in the document]
Comment (mcleinman) [reply qtglean/metrics.yaml]: Added, thanks.

File: src/apps/vpn/dnshelper.cpp [hunk 9]
```diff
@@ -19,10 +19,28 @@ namespace {
...
+// static
+QString DNSHelper::getDNSType() {
+  return getDNSDetails("test").dnsType;
+}
+
+//static
+dnsData DNSHelper::getDNSDetails(const QString& fallback) {
   if (!Feature::get(Feature::Feature_customDNS)->isSupported()) {
-    return fallback;
+    return dnsData{fallback, "NoCustomDNSAvailable"};
```
[the remark this answers is not in the document]
Comment (mcleinman) [reply src/apps/vpn/dnshelper.cpp:46]: Updated in yaml file, thanks.

File: src/apps/vpn/dnshelper.cpp [hunk 10]
```diff
@@ -32,7 +50,7 @@ QString DNSHelper::getDNS(const QString& fallback) {
 
   // Gateway DNS
   if (dnsProviderFlags == SettingsHolder::Gateway) {
-    return fallback;
+    return dnsData{fallback, "Default"};
```
[the remark this answers is not in the document]
Comment (mcleinman) [reply src/apps/vpn/dnshelper.cpp:56]: Good catch, thanks - the yaml file was outdated. It's been updated.

File: qtglean/metrics.yaml (hunk 4 above)
Comment (mcleinman) [reply qtglean/metrics.yaml]: It was part of the "ensure this gets updated in every PR so it doesn't get to production without being in the `setting` collection" plan. I've merged in `main`, so this no longer is changed in this PR.

**Comment by mcleinman** [comment]
> Same comment about adding unit tests: please do if possible. If not, file a follow up to do it.

Unit tests - This is covered by https://mozilla-hub.atlassian.net/browse/VPN-4454. I additionally blocked that ticket on VPN-4445, as that would be needed for this scenario.

File: src/apps/vpn/dnshelper.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (brizental) [reply src/apps/vpn/dnshelper.cpp]: > never appears in a metric

Is it guaranteed to never appear, or could it appear if something goes (terribly) wrong? If the latter, it needs to be documented.

[the remark this answers is not in the document]
Comment (mcleinman) [reply src/apps/vpn/dnshelper.cpp]: Guaranteed. No `dnsData` struct ever puts the actual DNS address in `dnsType`, for privacy. And this function only returns the `dnsType`.

[the remark this answers is not in the document]
Comment (brizental) [reply src/apps/vpn/dnshelper.cpp]: Hum, in that case let's rename it to "IMPOSSIBLE" instead of "Test" and add a little comment saying why it is impossible. Why does it even have a fallback if are never falling back though? Can we just remove the fallback then?

File: qtglean/metrics.yaml [hunk 11]
```diff
@@ -56,6 +56,28 @@ settings:
...
+  dns_type:
+    type: string
+    description: |
+      The type of DNS a user has active.
+      NoCustomDNSAvailable - Custom DNS app feature isn't available
+      Default - App's default DNS server
+      Custom - User has provided a valid IP address for their DNS
+      (BlockAds | BlockTrackers | BlockAdsAndTrackers
+      | BlockMalware | BlockMalwareAndAds | BlockMalwareAndTrackers
+      | BlockMalwareAndAdsAndTrackers) - Using app's privacy DNS option
```
Comment (brizental) [inline qtglean/metrics.yaml:68]: nit: You can use Markdown here. Maybe reformat it as a list.

File: src/apps/vpn/dnshelper.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (mcleinman) [reply src/apps/vpn/dnshelper.cpp]: The current version has a comment and renamed it to `pretendDNS`. I like `IMPOSSIBLE` even better, thanks. Updating that now.

File: qtglean/metrics.yaml (hunk 11 above)
Comment (mcleinman) [reply qtglean/metrics.yaml]: Thanks for that - I've updated this.

**Review by brizental — APPROVED** [review]
